src/filetree.rs

```diff
@@ -302,40 +353,82 @@ pub(crate) fn apply_diff(
+            for path in new_diff.removals.iter() {
+                destdir
+                    .remove_file_optional(path)
+                    .with_context(|| format!("removing {path}"))?;
+            }
+        }
+        // Ensure all of the new files are written persistently to disk
+        if !opts.skip_sync {
+            syncfs(destdir)?;
+        }
+        // skip if there are no changes or additions
+        if new_diff.additions.len() == 0 && new_diff.changes.len() == 0 {
+            continue;
+        }
+        let src_path = efipath.join(&path);
+        let tmp_path = efipath.join(format!("{}.tmp", path));
+        if tmp_path.exists() {
+            std::fs::remove_dir_all(&tmp_path)?;
+        }
+        if let Some(parent) = tmp_path.parent() {
+            destdir.ensure_dir_all(parent, DEFAULT_FILE_MODE)?;
+        }
+        // in case there is additional with subdir
+        destdir.ensure_dir_all(&src_path, DEFAULT_FILE_MODE)?;
+        // copy original dir to tmp
+        copy_dir(&src_path, &tmp_path)
+            .with_context(|| format!("copying {:?} to {:?}", src_path, tmp_path))?;
+        assert!(destdir.exists(&tmp_path)?);
+
+        // Write new and changed files
+        for pathstr in new_diff.additions.iter().chain(new_diff.changes.iter()) {
+            let path = Path::new(pathstr);
+            let pathtmp = tmp_for_path(path);
+            if let Some(parent) = pathtmp.parent() {
+                destdir.ensure_dir_all(parent, DEFAULT_FILE_MODE)?;
+            }
+            destdir.remove_file_optional(&pathtmp)?;
+            srcdir
+                .copy_file_at(path, destdir, &pathtmp)
+                .with_context(|| format!("writing {:?} to {:?}", path, pathtmp))?;
+        }
+        // do local exchange
+        log::trace!(
+            "doing local exchange for {} and {}",
+            tmp_path.display(),
+            src_path.display()
+        );
         destdir
-            .local_rename(&pathtmp, path)
-            .with_context(|| format!("renaming {path}"))?;
-    }
-    if !opts.skip_removals {
-        for path in diff.removals.iter() {
-            destdir
-                .remove_file_optional(path)
-                .with_context(|| format!("removing {path}"))?;
+            .local_exchange(&tmp_path, &src_path)
+            .with_context(|| format!("exchange for {:?} and {:?}", tmp_path, src_path))?;
+        // finally remove the temp dir
+        log::trace!("cleanup: {}", tmp_path.display());
+        std::fs::remove_dir_all(&tmp_path).context("clean up temp")?;
+        // A second full filesystem sync to narrow any races rather than
+        // waiting for writeback to kick in.
+        if !opts.skip_sync {
+            syncfs(destdir)?;
         }
     }
-    // A second full filesystem sync to narrow any races rather than
-    // waiting for writeback to kick in.
-    if !opts.skip_sync {
-        syncfs(destdir)?;
-    }
 
     Ok(())
 }
```

Comment:
I think what would really help build confidence here is a few new unit tests. A good way to implement this would be by e.g. checking the ctime (creation time) on files. If we did any extra copying we shouldn't have, then the ctime on the files will have changed when it shouldn't.

Reply:
Sorry that I changed the logic, use the first sub dir, if change like `fedora/subdir/foo`, will copy to `feodra.tmp/`, this might also copy other files that should not update. Not sure if the current change is OK.